example.py

```diff
@@ -0,0 +1,223 @@
+#         return g.op("com.microsoft::AllReduce", x).setType(x.type())
+
+
+class AllReduce(torch.nn.Module):
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        if torch.onnx.is_in_onnx_export():
+            return torch.onnx.ops.symbolic(
+                "com.microsoft::AllReduce",
+                (x,),
+                {},
+                dtype=x.dtype,
+                shape=x.shape,
+            )
+        torch.distributed.all_reduce(x, op=torch.distributed.ReduceOp.SUM)
+        return x
+
+
+# https://github.com/microsoft/Olive/blob/7df3c0353b3b754dec23faee2d73f22af9526155/olive/common/hf/quant.py#L122
+
+# class QuantLinearTorchFunction(torch.autograd.Function):
+#     """Used to export the quantized linear layer to onnx using the contrib operator MatMulNBits."""
+
+#     @staticmethod
+#     def symbolic(g, x, qweight, scales, qzeros, g_idx, bits, group_size, in_features, out_features):
+#         tensor_args = [x, qweight, scales, qzeros]
+#         if g_idx is not None:
+#             tensor_args.append(g_idx)
+#         attrs = {
+#             "K_i": in_features,
+#             "N_i": out_features,
+#             "bits_i": bits,
+#             "block_size_i": group_size,
+#         }
+
+#         output = g.op(
+#             "com.microsoft::MatMulNBits",
+#             *tensor_args,
+#             # what does this outputs do?
+#             outputs=1,
+#             **attrs,
+#         )
+#         input_shape = x.type().varyingSizes()
+#         if input_shape is not None and hasattr(x.type(), "with_sizes"):
+#             output_type = x.type().with_sizes(input_shape[:-1] + [qweight.type().varyingSizes()[0]])
+#             output.setType(output_type)
+
+#         return output
+
+#     @staticmethod
+#     def forward(ctx, x, qweight, scales, qzeros, g_idx, bits, group_size, in_features, out_features):
+#         if torch.onnx.is_in_onnx_export():
+#             return torch.zeros(x.shape[:-1] + (out_features,), dtype=x.dtype, device=x.device)
+#         raise NotImplementedError("QuantLinearTorchFunction forward is only implemented for onnx export")
+
+
+class QuantLinearTorchFunction(torch.nn.Module):
+    def forward(
+        self,
+        x: torch.Tensor,
+        qweight,
+        scales,
+        qzeros,
+        g_idx,
+        bits: int,
+        group_size: int,
+        in_features: int,
+        out_features: int,
+    ):
+        if torch.onnx.is_in_onnx_export():
+            tensor_args = [x, qweight, scales, qzeros]
+            if g_idx is not None:
+                tensor_args.append(g_idx)
+            return torch.onnx.ops.symbolic(
+                "com.microsoft::QuantLinear",
+                tensor_args,
+                dict(
+                    bits=bits,
+                    block_size=group_size,
+                    K=in_features,
+                    N=out_features,
+                ),
+                dtype=x.dtype,
+                shape=(*x.shape[:-1], out_features),
```

Comment:
Can users create new SymInts? And use them in the shape? What's the API for creating a new SymInt (if it is possible)?

Reply:
The symints are always linked to inputs. Users can take a slice from some input/intermediate tensors and use them here

Reply:
I thought they supported or planned to support ops like NonZero creating a tensor of a new dynamic size?

Reply:
I see. It may still interact with the internal shape env pytorch maintains. I need to look deeper.